### PR TITLE
Photon production fix

### PR DESCRIPTION
**Pull request by amandalund**
This PR introduces some fixes to the photon production implementation. The main change is accounting for summed reactions: photons produced in neutron interactions can't always be assigned to a particular reaction, and in this case they may be described using summed reactions such as MT=3, MT=4, and MT=103-107. Summed reactions were already stored in the `IncidentNeutron` object; now any that produce photons are written out to the HDF5 library and the logic for handling summed reactions is added on the Fortran/C++ side. In addition to the photon fixes, secondary angle/energy distributions were modified to correctly sample discrete energies. Overall the changes to the code are pretty minimal.

I've done some comparisons with MCNP6 on a broomstick problem, where a monoenergetic, monodirectional neutron source directed down a thin, infinitely long cylinder produces secondary particles which escape before they have a chance to interact. The photon current is tallied over the surface of the cylinder. The comparisons were done for a few different source energies and many different nuclides using ENDF/B-VII.1 ACE data from MCNP6. The results are looking good -- haven't checked every nuclide, but so far everything is consistent. I'll post some representative plots below, with the relative error (in green) plotted alongside the spectra.

al27-14 0mev_photon

pb208-14 0mev_photon

u235-1 0mev_photon

u238-1 0mev_photon

### CODE REVIEW

File: openmc/data/neutron.py [hunk 1]
```diff
@@ -480,6 +480,12 @@ def export_to_hdf5(self, path, mode='a', libver='earliest'):
                 tgroup = g.create_group('total_nu')
                 rx.derived_products[0].to_hdf5(tgroup)
 
+        # Write redundant reaction data only for reactions with photon production
+        for rx in self.redundant_reactions.values():
+            if any([p.particle == 'photon' for p in rx.products]):
```
Comment (paulromano) [inline openmc/data/neutron.py]: Passing a generator expression here would avoid the creation of an extra list:
```Python
if any(p.particle == 'photon' for p in rx.products):
    ...
```
Note that when a generator expression is the only argument in a function call, it's not necessary to put an extra set of parentheses around it

File: openmc/data/neutron.py [hunk 2]
```diff
@@ -671,19 +682,28 @@ def from_ace(cls, ace_or_filename, metastable_scheme='nndc'):
                          'cross section is given.'.format(mt))
                     continue
 
-                # Create summed reaction with appropriate cross section
+                # Don't duplicate fission cross sections
+                if mt == 18:
+                    continue
```
Comment (paulromano) [inline openmc/data/neutron.py]: Why is this necessary? I would have thought that if mt=18 was already present, the code wouldn't make it here by virtue of `if mt not in data` above.

Comment (amandalund) [reply openmc/data/neutron.py]: If MT=18 is *not* present for a fissionable material, then the partial fission MTs are. There were a few cases (e.g. U234 and U240) where we were double counting fission cross sections.

**Comment by paulromano** [comment]
Just wanted to put a note here for posterity -- @amandalund and I had a discussion offline about photon production for fission. Before, in cases where photon production was assigned to MT=18 in an ACE file but only partial fission reactions (MT=19,20,21,38) had neutron cross sections, we would just apply the MT=18 photon production to the individual reactions. With Amanda's changes here, that's no longer necessary since we can now arbitrarily store redundant reactions with photon production.
